### ISSUE TITLE
[BUG] Initializing a non-trivial `thread_local` struct before intializing rmm in PTDS causes a cuda error at exit

### ISSUE THREAD

**Issue by devavret** [rapidsai/rmm#1371]
**Describe the bug**
When using RMM in a target built with PTDS, initializing a `thread_local` struct with a non-default destructor before initializing stream ordered resource causes a cuda error at exit.

**Steps/Code to reproduce bug**
Apply the following changes to the pool mr tests and run `POOL_MR_PTDS_TEST`.
```diff
diff --git a/tests/mr/device/pool_mr_tests.cpp b/tests/mr/device/pool_mr_tests.cpp
index 4a234d2f..63b15f41 100644
--- a/tests/mr/device/pool_mr_tests.cpp
+++ b/tests/mr/device/pool_mr_tests.cpp
@@ -37,6 +37,32 @@ using cuda_mr     = rmm::mr::cuda_memory_resource;
 using pool_mr     = rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>;
 using limiting_mr = rmm::mr::limiting_resource_adaptor<rmm::mr::cuda_memory_resource>;
 
+class Stack {
+ public:
+  Stack() noexcept : size_(0), capacity_(0), base_(nullptr) {}
+
+  // ~Stack() noexcept { delete[] base_; }
+
+  // Breaks even with empty destructor
+  ~Stack() noexcept {}
+
+  // Works when defaulted
+  // ~Stack() noexcept = default;
+
+  size_t size_;
+  size_t capacity_;
+  int* base_;
+};
+
+static Stack& GetStack()
+{
+  // Crashes if thread_local. Works if only static
+  thread_local Stack stack_ = Stack();
+  return stack_;
+}
+
+TEST(Another, ThreadLocal) { GetStack(); }
+
 TEST(PoolTest, ThrowOnNullUpstream)
 {
   auto construct_nullptr = []() { pool_mr mr{nullptr}; };
```
The test runs to completion but segfaults at the end with the following cuda error.
```
[  PASSED  ] 11 tests.
CUDA Error detected. cudaErrorCudartUnloading driver shutting down
POOL_MR_PTDS_TEST: /home/devavret/Development/rapids/rmm/include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:270: rmm::mr::detail::stream_ordered_memory_resource<PoolResource, FreeListType>::event_wrapper::~event_wrapper() [with PoolResource = rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>; FreeListType = rmm::mr::detail::coalescing_free_list]: Assertion `status__ == cudaSuccess' failed.
Aborted (core dumped)
```

**Expected behavior**
No CUDA error should be thrown.

**Additional Information**
This is more reliably reproducible with a debug build.

**Comment by harrism** [comment]
Hi @devavret, thanks for the issue!

I tried adding your code locally but I don't get any errors. 

```
(rapids) coder ➜ ~/rmm/build/release $ gtests/POOL_MR_PTDS_TEST 
Running main() from gmock_main.cc
[==========] Running 12 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 1 test from Another
[ RUN      ] Another.ThreadLocal
[       OK ] Another.ThreadLocal (0 ms)
[----------] 1 test from Another (0 ms total)

[----------] 11 tests from PoolTest
[ RUN      ] PoolTest.ThrowOnNullUpstream
[       OK ] PoolTest.ThrowOnNullUpstream (31 ms)
[ RUN      ] PoolTest.ThrowMaxLessThanInitial
[       OK ] PoolTest.ThrowMaxLessThanInitial (0 ms)
[ RUN      ] PoolTest.ReferenceThrowMaxLessThanInitial
[       OK ] PoolTest.ReferenceThrowMaxLessThanInitial (0 ms)
[ RUN      ] PoolTest.AllocateNinetyPercent
[       OK ] PoolTest.AllocateNinetyPercent (90 ms)
[ RUN      ] PoolTest.TwoLargeBuffers
[       OK ] PoolTest.TwoLargeBuffers (901 ms)
[ RUN      ] PoolTest.ForceGrowth
[       OK ] PoolTest.ForceGrowth (4 ms)
[ RUN      ] PoolTest.DeletedStream
[       OK ] PoolTest.DeletedStream (2 ms)
[ RUN      ] PoolTest.InitialAndMaxPoolSizeEqual
[       OK ] PoolTest.InitialAndMaxPoolSizeEqual (2 ms)
[ RUN      ] PoolTest.NonAlignedPoolSize
[       OK ] PoolTest.NonAlignedPoolSize (0 ms)
[ RUN      ] PoolTest.UpstreamDoesntSupportMemInfo
[       OK ] PoolTest.UpstreamDoesntSupportMemInfo (508 ms)
[ RUN      ] PoolTest.MultidevicePool
[       OK ] PoolTest.MultidevicePool (0 ms)
[----------] 11 tests from PoolTest (1540 ms total)

[----------] Global test environment tear-down
[==========] 12 tests from 2 test suites ran. (1540 ms total)
[  PASSED  ] 12 tests.
```
Can you provide more info about your setup? I'm on RMM `branch-23.12` running in the RMM dev container `cuda12.0-conda`. I tested on a GV100 with driver `Driver Version: 535.104.05`

**Comment by devavret** [comment]
Im on rmm branch 23.08 and I'm using cuda 11.7 with driver 525.147.05 on a RTX 3050Ti mobile. 

Can you try with a debug build?

**Comment by wence-** [comment]
I think this could be https://github.com/rapidsai/rmm/issues/1306. There were some issues with the thread_local storage in the stream ordered allocator which meant we were in some circumstances recording on events with the wrong device/stream pair active. This would exhibit as an error somewhere in the event deallocation. _But_ since a destructor can't throw, the cuda error checking in the destructor doesn't run unless you're compiling in debug mode.

**Comment by wence-** [comment]
Hmm, that was not it, but I can repro with 23.10 and compiling with debug enabled.

**Comment by wence-** [comment]
In a rapids-compose environment (providing cudart-12.0) I can produce the same error with (no RMM involved, no need for PTDS):

```c++
#include <cassert>
#include <cuda_runtime_api.h>
#include <iostream>

class Stack {
public:
  // Breaks even with empty destructor
  ~Stack() noexcept {}
};

struct event_wrapper {
  event_wrapper() {
    auto err = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
    assert(err == cudaSuccess);
  }
  ~event_wrapper() {
    auto err = cudaEventDestroy(event);
    std::cout << err << std::endl;
    assert(err == cudaSuccess);
  }
  cudaEvent_t event{};

  event_wrapper(event_wrapper const &) = delete;
  event_wrapper &operator=(event_wrapper const &) = delete;
  event_wrapper(event_wrapper &&) noexcept = delete;
  event_wrapper &operator=(event_wrapper &&) = delete;
};

void get_event() {
  thread_local auto ew = event_wrapper{};
  return;
}
int main(void) {
  // [[maybe_unused]] auto x = cudaFree(0);
  [[maybe_unused]] thread_local auto stack = Stack();
  get_event();
  return 0;
}
```

The destructor for `event_wrapper` is run by the exit handlers (below main). My suspicion is that the existence of the `Stack` destructor running during thread local atexit handling somehow messes up ordering of the destructors such that the driver shutdown runs before the destructor for `event_wrapper` (I am guessing that the cuda runtime initialises the driver in such a way that that it sets up an atexit shutdown handler).

Note that if I (implicitly) initialise a context and the driver by calling `cudaFree(0)` _before_ initialising the thread-local `Stack` object, I do not get an error.

Interestingly, I can't reproduce the problem on bare-metal with either CTK-11.8 or CTK-12.3.

**Comment by wence-** [comment]
tl;dr: Depending on how your libstdc++ was compiled, you need to arrange to call `cudaFree(0)` (to initialise the runtime) _before_ any objects with `thread_local` storage are initialised.

**Details:**

Having done a bit more debugging, I _think_ I know what is going on (and why I couldn't reproduce on bare metal).

`thread_local` destructors are run in the libc's `atexit` handler (I assume glibc...). Depending on which libstdc++ you have, the way this is done is in one of two ways:

https://github.com/bminor/glibc/blob/5dd3bda59c2d9da138f0d98808d087cdb95cdc17/stdlib/exit.c#L36

Depending on how your libstdc++ was built, it either registers `thread_local` dtors to be called via `__call_tls_dtors` 

https://github.com/gcc-mirror/gcc/blob/094a609c69e5608311a39ff66eb646ca3be70652/libstdc%2B%2B-v3/libsupc%2B%2B/atexit_thread.cc#L58 

or else has its own cleanup routine:

https://github.com/gcc-mirror/gcc/blob/094a609c69e5608311a39ff66eb646ca3be70652/libstdc%2B%2B-v3/libsupc%2B%2B/atexit_thread.cc#L134-L137

Note the comment: 
```
    // Also make sure the destructors are run by std::exit.
    // FIXME TLS cleanups should run before static cleanups and atexit
    // cleanups.
```

OK, so if thread local storage dtors are run via `__call_tls_dtors`, everything is fine (these are guaranteed to run before any other atexit handlers).

However, if this is _not_ the case, thread local storage dtors are run in an order (relative to other atexit handlers) that depends on when the _first_ `thread_local` object is initialised.

AFAICT (I haven't look at the cuda runtime source), the implicit initialisation of the driver that happens when you first call a cuda runtime API function registers something to run `atexit` that arranges to unload the driver. So, what I think is happening is:

atexit handlers are run last-in, first-out.

**Bad case**

```
(1) thread_local auto stack = Stack{};
(2) thread_local auto ew = Event{};
```

(1) registers the libstdc++ atexit handler that tears down all the thread local objects by calling their dtors,
(2) calls into the cuda runtime which loads the driver, and registers an atexit handler that unloads the driver.

So after this, the list of atexit handlers is:

```
[libstdc++_atexit, // will call ~Event(), then ~Stack()
 cudart_unload_driver]
```

Since we're LIFO, we first unload the driver, then call `~Stack`, then call `~Event`. But `~Event` wants to call into the runtime, hence we get an error.

In contrast,

**Good case**

If your (gnu) libstdc++ was compiled with `_GLIBCXX_HAVE___CXA_THREAD_ATEXIT` then the `thread_local` storage dtors are registered in glibc's `tls_dtor_list` which is cleared out before any explicitly registered `atexit` handlers have run. So we do: `~Stack()`, `~Event()`, `cudart_unload_driver()` and all is right with the world.

**Mitigation**

Arrange to initialise the runtime (and hence any the driver-unloading atexit handler) _before_ any `thread_local` storage is initialised (e.g. by calling `cudaFree(0)`). I'm not sure if there is a more principled way of doing this.

**Comment by wence-** [comment]
I checked and it is considered undefined behaviour to call into the runtime below main (eg in atexit) which is what the RMM thread_local storage that destructs the events does. So we will need to figure out a way to address this in RMM.

**Comment by jrhemstad** [comment]
> So we will need to figure out a way to address this in RMM.

Just don't destroy it. Just do:
```
auto* ptr = new rmm::cuda_event{}
```
and never free it. The OS will reclaim the object.

**Comment by harrism** [comment]
Wow, epic debugging @wence-! 🚀

**Comment by harrism** [comment]
@jrhemstad (there's no such thing as `rmm::cuda_event`) this is the code that creates TLS:

```
thread_local std::vector<std::shared_ptr<event_wrapper>> events_tls(
        rmm::get_num_cuda_devices());
```

There are lots of hacky ways I could avoid calling the destructor of `event_wrapper`. The `new` approach you mention above is not so easy to implement since it's a vector.  I could create a separate wrapper struct with a default destructor (so it doesn't destroy the events). Or add a bool template argument to the existing one that is used to decide whether or not to destroy the event.

This all feels hacky especially given the comment above that makes it sound like this ordering of destruction of TLS is a bug. 

```
// Also make sure the destructors are run by std::exit.
// FIXME TLS cleanups should run before static cleanups and atexit
// cleanups.
```

**Comment by wence-** [comment]
I have a PoC fixing this at https://github.com/rapidsai/rmm/pull/1375, it removes the `event_wrapper` and just stores (and leaks) raw `cudaEvent_t` objects. The `event_wrapper` is only used for the PTDS thread_local case anyway, which is why this works.

wrt the TLS ordering comment, having read the standard in more detail, I believe that although that implementation is wrong, the FIXME in the bug is _also wrong_.

[`basic.term.start#5`](https://eel.is/c++draft/basic.start.term#5), emphasis mine:

> If the completion of the initialization of an object with **static storage duration strongly happens before a call to std​::​atexit** (see [<cstdlib>](https://eel.is/c++draft/cstdlib.syn#header:%3ccstdlib%3e), [[support.start.term]](https://eel.is/c++draft/support.start.term)), the call to **the function passed to std​::​atexit is sequenced before the call to the destructor for the object**. [...]

Since objects annotated with `thread_local` at block scope have static storage duration, calls to their destructors should be sequenced appropriately with calls to the functions registered with `atexit`:

Consider:
```c++
#include <cstdlib>
#include <iostream>
struct A {
  ~A() { std::cout << "~A" << std::endl; }
};

struct B {
  ~B() {
    std::cout << "~B" << std::endl;
  }
};

#ifndef DCL
#define DCL
#endif
int main()
{
  DCL A a{};
  std::atexit([]() { std::cout << "~atexit" << std::endl; });
  DCL B b{};
  return 0;
}
```

```
$ c++ -std=c++11 bug.cxx -o normal
$ c++ -std=c++11 bug.cxx -DDCL=static -o static
$ c++ -std=c++11 bug.cxx -DDCL=thread_local -o thread_local
$ ./normal
# what we expect
~B
~A
~atexit
$ ./static
# atexit registration strongly happens-before initialisation of B, so ~B is sequenced before the atexit callback, as expected
~B
~atexit
~A
$ ./thread_local
# This, I think, should be identical to the static case, because the ordering is the same.
# if using LLVM's libc++ or gnu libstdc++ compiled with _GLIBCXX_HAVE___CXA_THREAD_ATEXIT
~B
~A
~atexit
# Or, if using gnu libstdc++ compiled without _GLIBCXX_HAVE___CXA_THREAD_ATEXIT
~atexit
~B
~A
```

I contend that _both_ of the results with `thread_local` are wrong.
